### PR TITLE
Tag Page Cleanup

### PR DESCRIPTION
**Pull request by arthurtyukayev**
The previous tag page was a bit messy that and was missing a few conventions that would help the project in the future as it. I propose a refactor throughout the project to match some of the pattern that I've created for the Tags Page.

Most importantly, the components folder should contain sub folders that match the pages. These folders should contain the components that are used within the page. Also, it's important to keep the Route / Page component render methods small, which will improve the ability to test in the future. And finally, we need tests.

### CODE REVIEW

**Comment by raiyni** [comment]
I disagree. It's a 1 page route. It doesn't need to be broken into components for no reason.

**Comment by arthurtyukayev** [comment]
The Tags page itself is sort of an odd example, as it's technically a two page route, that should be a one page route. The project splits up the route into a `tag` and `tag-show` and actually does an entire route change for something that really should just be a single page. 

Breaking the page out into components will help with testing in the future, as more features are added to the tags feature, keeping it in the same file will make this file difficult to maintain.

**Comment by deathbeam** [comment]
It is not single page so people can share the URL with tags. What is the purpose of splitting tag page to components when the components will be only ever used in tag page? Also, whats up with the utility meta function, I already created <Meta> component

**Comment by arthurtyukayev** [comment]
It can be a single page and still shareable. That's the whole point of a SPA. The point of splitting the pages into components is to allow for easier unit testing, as well as easier styling.
As for the `Meta` component, it's not really being used correctly, it doesn't mount anything. It's rather unclear what it does. The page constructor is a better place to set the page title.

As for the Tag and Tag Show, I can update this PR with an example of what it would look like if we combined the two pages. React is designed for this purpose, and because page is nearly the same, there is no reason to break SPA to create this functionality.

**Comment by deathbeam** [comment]
Not really, I can't share something with someone when I dont have unique URL. The landing tag page simply redirects you to the respective URL. And as those 2 pages look completely different there is no reason to have them grouped into 1. 

But page constructor is incorrect place for it to be set, sometimes you dont have data in constructor that you need, and sometimes the page is not recreated, just props are changed (e.g see XP tracker page) and so constructor will not even be called multiple times. I would rather have it unified than having special cases like this.

It works as replacement for `react-head` (or w/e is it called) and works in very similar way, just more lightweight. It does not matter if it actually mounts anything to DOM or nah.

As I said before, I do not see a reason why the pages should be unified into 1 as they are completely different.

**Comment by arthurtyukayev** [comment]
It's possible to simply update the URL when hitting "Go" and the page can be updated to support both states. Either way, I can see how splitting it up could have been a design decision. I think that it would function better as a single page, it would allow people to update their tags without having to navigate back and forth between pages, it's really a usability update, but I can see the validity of making it two pages.

For the meta updates, if we need updates then moving it to `component(Will|Did)Update()` should work just as well. That's essentially what the Meta component is doing, but it's got a bunch of overhead because it's a component. I don't think there really needs to be a component in the render that manages the head, it seems like an odd pattern to me, especially when it could just be done with a simple utility function.

File: src/components/app.js [hunk 1]
```diff
@@ -2,15 +2,16 @@ import '@fortawesome/fontawesome-free/css/all.min.css'
 import { h } from 'preact'
 import { connect } from 'preact-redux'
 import Router from 'preact-router'
+
```
Comment (deathbeam) [inline src/components/app.js]: why is this newline here?

File: src/components/app.js [hunk 2]
```diff
@@ -46,7 +47,9 @@ const App = ({ loading, stars, navbarDark }) => (
   </div>
 )
 
-export default connect(state => ({
+const mapStateToProps = state => ({
```
Comment (deathbeam) [inline src/components/app.js]: any reason why this function should be named?

File: src/components/tag/Form.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import { h } from 'preact'
+
+const TagForm = ({ setActiveTag, activeTag }) => (
```
Comment (deathbeam) [inline src/components/tag/Form.js]: I would make this more generic, and not related to tags at all (and also filename casing is wrong)

File: src/routes/tag.js [hunk 4]
```diff
@@ -1,43 +1,45 @@
...
-import { connect } from 'preact-redux'
+import { setMetaDetails } from '../utilities/meta-utils'
+import hero from '../_data/hero'
+import Layout from '../components/layout'
+import TagForm from '../components/tag/form'
 
 class Tag extends Component {
-  render({ activeTag, setActiveTag }) {
+  componentDidUpdate() {
+    setMetaDetails({ title: `Select tag tab - ${hero.title}` })
```
Comment (deathbeam) [inline src/routes/tag.js]: As I said before, remove this setMetaDetails function, I already have <Meta> component

File: src/routes/tag.js [hunk 5]
```diff
@@ -1,43 +1,45 @@
...
     return (
       <div>
-        <Meta title={`Select tag tab - ${hero.title}`} />
         <Layout class="tag-container">
           <h5>
             Enter tag <br />
             <small class="text-muted">
-              Paste a tag exported from Runelite's Bank Tags plugin and click
-              'GO'
+              {
```
Comment (deathbeam) [inline src/routes/tag.js]: Any reason why you put this inside {}? It was perfectly valid before

File: src/routes/tag.js [hunk 6]
```diff
@@ -1,43 +1,45 @@
...
-            Go
-          </a>
+          <TagForm setActiveTag={setActiveTag} activeTag={activeTag} />
         </Layout>
       </div>
     )
   }
 }
 
+const mapStateToProps = ({ tag }) => ({ activeTag: tag.activeTag })
```
Comment (deathbeam) [inline src/routes/tag.js]: And again, any reason why you are naming this functions

**Comment by arthurtyukayev** [comment]
Closing, will re-open when I had a better refactor done.

**Pull request closed without merging**